enterprise/server/ociregistry/ociregistry.go

```diff
@@ -0,0 +1,305 @@
+package ociregistry
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/environment"
+	"github.com/buildbuddy-io/buildbuddy/server/metrics"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/buildbuddy-io/buildbuddy/server/util/prefix"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+)
+
+const (
+	rangeHeaderBytesPrefix = "bytes="
+
+	registryInstanceName = "registry_artifacts"
+)
+
+var (
+	manifestReqRE = regexp.MustCompile("/v2/(.+?)/manifests/(.+)")
+	blobReqRE     = regexp.MustCompile("/v2/(.+?)/blobs/(.+)")
+	blobUploadRE  = regexp.MustCompile("/v2/(.+?)/blobs/uploads/")
+
+	enableRegistry = flag.Bool("ociregistry.enabled", false, "Whether to enable registry services")
+)
+
+type registry struct {
+	env environment.Env
+}
+
+func Register(env environment.Env) error {
+	if !*enableRegistry {
+		return nil
+	}
+
+	mux := env.GetInternalHTTPMux()
+	if mux == nil {
+		return status.FailedPreconditionErrorf("Registry requires internal HTTP mux")
+	}
+
+	r, err := New(env)
+	if err != nil {
+		return err
+	}
+
+	mux.Handle("/v2/", r)
+	return nil
+}
+
+func New(env environment.Env) (*registry, error) {
+	r := &registry{
+		env: env,
+	}
+	return r, nil
+}
+
+func (r *registry) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	r.handleRegistryRequest(w, req)
+}
+
+func (r *registry) handleRegistryRequest(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	ctx, err := prefix.AttachUserPrefixToContext(ctx, r.env)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("could not attach user prefix: %s", err), http.StatusInternalServerError)
+		return
+	}
+	authHeader := req.Header.Get("Authorization")
+	log.CtxInfof(ctx, "%s %q '%s'", req.Method, req.RequestURI, authHeader)
+	// Clients issue a GET /v2/ request to verify that this  is a registry
+	// endpoint.
+	if req.RequestURI == "/v2/" {
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+	forwardedRegistry := req.Header.Get("X-Forwarded-Host")
```

Comment:
might be nice to leave a comment somewhere about these headers. I found

//   - The X-Forwarded-For header is set to the client IP address.
//   - The X-Forwarded-Host header is set to the host name requested
//     by the client.
//   - The X-Forwarded-Proto header is set to "http" or "https", depending
//     on whether the inbound request was made on a TLS-enabled connection.